src/slide-image-source.cc

```diff
@@ -0,0 +1,164 @@
+// (c) 2021 Leonardo Romor <leonardo.romor@gmail.com>,
+//          Henner Zeller <h.zeller@acm.org>
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation version 2.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+#include "slide-image-source.h"
+
+#include "framebuffer.h"
+#include "terminal-canvas.h"
+
+#include <cassert>
+#include <cstdio>
+#include <unistd.h>
+
+#include <limits>
+#include <string>
+#include <utility>
+
+#include <openslide.h>
+
+extern "C" {
+#  include <libswscale/swscale.h>
+}
+
+namespace timg {
+const char *SlideImageSource::VersionInfo() {
+    return openslide_get_version();
+}
+
+static void dummy_log(void *, int, const char *, va_list) {}
+
+// Scoped clean-up of c objects.
+struct ScopeGuard {
+    ScopeGuard(std::function<void()> f) : f_(f) {}
+    ~ScopeGuard() { f_(); }
+    std::function<void()> f_;
+};
+
+static inline bool invalid_dimensions(int64_t width, int64_t height) {
+    return (width < 0 || height < 0) ||
+        (width > std::numeric_limits<int>::max() ||
+         height > std::numeric_limits<int>::max());
+}
+
+bool SlideImageSource::LoadAndScale(const DisplayOptions &opts, int, int) {
+    options_ = opts;
+    if (opts.scroll_animation ||
+        filename() == "/dev/stdin" || filename() == "-") {
+        return false;  // Not dealing with these now.
+    }
+
+    // Check if this format is supported
+    openslide_t *osr = openslide_open(filename().c_str());
+    if (!osr)
+        return false;
+
+    // Automatically cleanup openslide context
+    ScopeGuard s([osr]() { openslide_close(osr); });
+
+    std::unique_ptr<timg::Framebuffer> source_image;
+    int64_t width, height, l0_width, l0_height;
+    int target_width;
+    int target_height;
+
+    // We need to use one of the layers of the pyramid.
+    // Get native the highest resolution image.
+    openslide_get_level0_dimensions(osr, &l0_width, &l0_height);
+    if (invalid_dimensions(l0_width, l0_height))
+        return false;
+
+    // Get the target size
+    CalcScaleToFitDisplay(l0_width, l0_height, opts, false,
+                          &target_width, &target_height);
+
+    // First check if there's a thumbnail with enough resolution.
+    const char* const* associated_image_names = openslide_get_associated_image_names(osr);
+    while (*associated_image_names) {
+        const char *name = *associated_image_names;
+        openslide_get_associated_image_dimensions(osr, name, &width, &height);
+        if (invalid_dimensions(width, height))
```

Comment:
Could this be a `continue` reason, or do we know that at that point things are broken, so we have to `return false` ?

Reply:
I think the second. If openslide is returning a dim < 0 that's an error. Every associated image needs to have a valid dimension. The second check about limits is to be sure that I can interoperate between openslide dimensions (int64_t) and your function dimension type (int).
(solved)